javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -199,38 +199,65 @@ object ConcurrentHashMap {
   }
 
   /* `KeySetView` is a public class in the JDK API. The result of
-   * `ConcurrentHashMap.keySet()` must be statically typed as a `KeySetView`,
-   * hence the existence of this class, although it forwards all its operations
-   * to the inner key set.
+   * `ConcurrentHashMap.keySet()` must be statically typed as a `KeySetView`.
    */
-  class KeySetView[K, V] private[ConcurrentHashMap] (inner: Set[K])
+  class KeySetView[K, V] private[ConcurrentHashMap] (innerMap: InnerHashMap[K, V], defaultValue: V)
       extends Set[K] with Serializable {
 
-    def contains(o: Any): Boolean = inner.contains(o)
+    def contains(o: Any): Boolean = innerMap.containsKey(o)
 
-    def remove(o: Any): Boolean = inner.remove(o)
+    def remove(o: Any): Boolean = innerMap.remove(o) != null
 
-    def iterator(): Iterator[K] = inner.iterator()
+    def iterator(): Iterator[K] = innerMap.keySet().iterator()
 
-    def size(): Int = inner.size()
+    def size(): Int = innerMap.size()
 
-    def isEmpty(): Boolean = inner.isEmpty()
+    def isEmpty(): Boolean = innerMap.isEmpty()
 
-    def toArray(): Array[AnyRef] = inner.toArray()
+    def toArray(): Array[AnyRef] = innerMap.keySet().toArray()
 
-    def toArray[T <: AnyRef](a: Array[T]): Array[T] = inner.toArray[T](a)
+    def toArray[T <: AnyRef](a: Array[T]): Array[T] = innerMap.keySet().toArray(a)
 
-    def add(e: K): Boolean = inner.add(e)
+    def add(e: K): Boolean = {
+      if (e == null || defaultValue == null) {
```

Comment:
I'm pretty positive the `null` check here is wrong. If `c` is `null`, this should throw an `NPE`, not an `UnsupportedOperationException`.

I'd just leave the `null` check to `putIfAbsent`.

Reply:
Hmm, but removing `e == null` fails on `SetTest#contains`.
It seems that `putIfAbsent` on InnerHashMap throw nothing on `null` key.

Reply:
Hah, indeed. It seems like `ConcurrentHashMap` explicitly fixes this behavior. Was this an oversight in f7be410503f87b402a86c92945dba37455f06853? @sjrd ? If yes, maybe this should be moved to `NullRejectingHashMap`.

Reply:
Ah yes, well spotted. Indeed, introducing the optimized implementation of `putIfAbsent` in `HashMap` bypassed the null checks that `NullRejectingHashMap` performs. So indeed, we should override `putIfAbsent` in `NullRejectingHashMap` to perform the null checks before calling `super.putIfAbsent` (similarly to `put`).

Reply:
Fixed in https://github.com/scala-js/scala-js/pull/4367/commits/16687e688329b3ed7c612f33e030517278d10a2a

Reply:
Hum, did this line also change? `e == null` here is definitely wrong. If `e == null`, this should throw an NPE (but `innerMap.putIfAbsent` will already do so).